src/Caching/StackExchangeRedis/src/StackExchangeRedisCacheServiceCollectionExtensions.cs

```diff
@@ -32,8 +34,16 @@ public static IServiceCollection AddStackExchangeRedisCache(this IServiceCollect
         }
 
         services.AddOptions();
+
+#if NET7_0_OR_GREATER
+        services.AddLogging();
+#else
+        services.TryAddSingleton<ILoggerFactory>(Logging.Abstractions.NullLoggerFactory.Instance);
+        services.TryAdd(ServiceDescriptor.Singleton(typeof(ILogger<>), typeof(Logger<>)));
+#endif
+
         services.Configure(setupAction);
-        services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCache>());
+        services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCacheImpl>());
```

Comment:
Is switching the service descriptor to a factory really any better than switching the implementation type?

The runtime type of the object the factory returns would still be `RedisCacheImpl`. And `ImplementationFactory` is declared as just a `Func<IServiceProvider, object>` at compile time. The runtime type of the func is different because of covariance, but who's looking at the runtime type of `ImplementationFactory`? Not the DI system itself. I'd stick with the type over the factory because the DI can optimize it better.

I don't remember anyone complaining about the `KestrelServer` and `KestrelServerImpl` split, and that was even more breaking thinking back. If someone was calling `(KestrelServer)app.Services.GetRequiredService<IServer>()`, it would have broke. At least `(RedisCache)app.Services.GetRequiredService<IDistributedCache>()` still works. I don't think we even made a breaking change announcement for the Kestrel change.

Reply:
The breaking part is doing `app.Services.GetRequiredService<RedisCache>()`, which is not a common pattern probably, but still breaking.

Reply:
> `app.Services.GetRequiredService<RedisCache>()`

I don't think this ever worked if you just called `builder.Services.AddStackExchangeRedisCache()`. If you called, `builder.Services.AddSingleton<RedisCache>()`, it would work. But that still works after this change. Now the only downside is that you don't get logging that way.

Reply:
I just tested this locally, and using `services.Add(ServiceDescriptor.Singleton<IDistributedCache, RedisCache>())` and then doing the below fails:

```csharp
[Fact]
public void AddStackExchangeRedisCache_RegistersRedisCache()
{
    // Arrange
    var services = new ServiceCollection();

    // Act
    services.AddStackExchangeRedisCache(options => { });

    // Assert
    var serviceProvider = services.BuildServiceProvider();
    var redisCache = serviceProvider.GetRequiredService<RedisCache>();

    Assert.NotNull(redisCache);
}
```

``` Microsoft.Extensions.Caching.StackExchangeRedis.CacheServiceExtensionsTests.AddStackExchangeRedisCache_RegistersRedisCache
   Source: CacheServiceExtensionsTests.cs line 17
   Duration: 11 ms

  Message: 
System.InvalidOperationException : No service for type 'Microsoft.Extensions.Caching.StackExchangeRedis.RedisCache' has been registered.

  Stack Trace: 
ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
CacheServiceExtensionsTests.AddStackExchangeRedisCache_RegistersRedisCache() line 27
```